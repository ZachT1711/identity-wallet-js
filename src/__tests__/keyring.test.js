const Keyring = require('../keyring')
const HDNode = require('@ethersproject/hdnode')


describe('Keyring', () => {

  let keyring1
  let keyring2
  let keyring3
  const seed = HDNode.mnemonicToSeed('clay rubber drama brush salute cream nerve wear stuff sentence trade conduct')

  it('throws error if no seed', async () => {
    expect(() => new Keyring()).toThrow()
  })

  it('derives correct keys from entropy', async () => {
    keyring2 = new Keyring('0xf0e4c2f76c58916ec258f246851bea091d14d4247a2fc3e18694461b1816e13b')
    keyring3 = new Keyring('0x24a0bc3a2a1d1404c0ab24bef9bb0618938ee892fbf62f63f82f015eddf1729e')
    expect(keyring2._seed).toEqual('0xf0e4c2f76c58916ec258f246851bea091d14d4247a2fc3e18694461b1816e13b')
  })

  it('derives correct keys from seed', async () => {
    keyring1 = new Keyring(seed)

    expect(keyring1.getPublicKeys()).toMatchSnapshot()
    expect(keyring1.getPublicKeys({ mgmtPub: true })).toMatchSnapshot()
    expect(keyring1.getPublicKeys({ uncompressed: true })).toMatchSnapshot()
    expect(keyring1.serialize()).toEqual(seed)
  })

  it('signs data correctly', async () => {
    expect((await keyring1.getJWTSigner()('asdf'))).toMatchSnapshot()
    expect((await keyring1.getJWTSigner('space1')('asdf'))).toMatchSnapshot()
    expect((await keyring1.getJWTSigner('space2')('asdf'))).toMatchSnapshot()
  })

  it('encrypts and decrypts correctly', () => {
    const testMsg = "Very secret test message"
    let box = keyring1.asymEncrypt(testMsg, keyring2.getPublicKeys().asymEncryptionKey)

    let cleartext = keyring2.asymDecrypt(box.ciphertext, box.ephemeralFrom, box.nonce)
    expect(cleartext).toEqual(testMsg)
  })

  it('symmetrically encrypts correctly', async () => {
    const testMsg = "Very secret test message"
    const box = keyring2.symEncrypt(testMsg)
    const cleartext = keyring2.symDecrypt(box.ciphertext, box.nonce)
    expect(cleartext).toEqual(testMsg)
  })

  it('encrypts and decrypts correctly with authSecret', () => {
    const testMsg = "Very secret test message"
    const authSecret = Buffer.from(Keyring.naclRandom(32)).toString('hex')

    const box = Keyring.encryptWithAuthSecret(testMsg, authSecret)
    const cleartext = Keyring.decryptWithAuthSecret(box.ciphertext, box.nonce, authSecret)
    expect(cleartext).toEqual(testMsg)
  })
})
