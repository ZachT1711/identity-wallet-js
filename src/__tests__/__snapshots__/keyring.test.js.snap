// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyring derives correct keys from seed 1`] = `
Object {
  "asymEncryptionKey": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
  "managementKey": "0xa7b76379659b6c19E4DFf46e0DfEa48AFEBA8fe3",
  "signingKey": "027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229",
}
`;

exports[`Keyring derives correct keys from seed 2`] = `
Object {
  "asymEncryptionKey": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
  "managementKey": "0350966498e802a007513d518dc12a0a0ab1e48addcec1beaa989c6ff5f5ed226c",
  "signingKey": "027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229",
}
`;

exports[`Keyring derives correct keys from seed 3`] = `
Object {
  "asymEncryptionKey": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
  "managementKey": "0xa7b76379659b6c19E4DFf46e0DfEa48AFEBA8fe3",
  "signingKey": "047ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa2298f6eda9fabcee1df8fbe7d49f7f8c14b736d05c4c8f4f276f1a523371ae202fa",
}
`;

exports[`Keyring signs data correctly 1`] = `
Object {
  "r": "c896c3198ae31e3038bb22bf1a7da49c3f3a77a5e8a41fc19cdb275532583205",
  "recoveryParam": 0,
  "s": "bd1387127e4a31b89ac8e103be6d7939c26adb13b6450968b10049f662b407ce",
}
`;

exports[`Keyring signs data correctly 2`] = `
Object {
  "r": "244a23056eba642037ab0606f08ce5a077ab6041e92977a9f65727a6be1b3497",
  "recoveryParam": 1,
  "s": "fb984e09a3eb1358fea41858e1fbfd3b453e1d8204f800ddf53ee81eaf0d30f6",
}
`;

exports[`Keyring signs data correctly 3`] = `
Object {
  "r": "b04fb687a16254cff30acee7f41fe3bf5e5d1d583c34ab869abae46e2dafa16c",
  "recoveryParam": 0,
  "s": "15d604fe58f44d349591ed36fb386f7b211c8e337bb115651d64199268ef752b",
}
`;
